# Task 4.7: Deriving new variables

## This script contains the following points:

### 1.Importing libraries

### 2. Importing data

### 3. Creating new variables - price_label and busiest_day

### 4. Create busiest_days variable

### 5. Create busiest_period_of_day variable

### 6. Export data

#### 1. Import libraries

In [42]:
# Import libraries

import pandas as pd
import numpy as np
import os

#### 2. Import data

In [43]:
# Import data

In [44]:
# Create path

path = r'C:\Users\Lenad\Documents\Data Analytics Immersion\Achievement 4\Jupyter folder\Instacart basket analysis'

In [45]:
# Import ords_prods_merge

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [46]:
# Create subset of dataframe showing only first 1,000,000 rows

df = df_ords_prods_merge[:1000000]

In [47]:
df.shape

(1000000, 15)

#### 3. Creating new variables - price_label and busiest_day

In [48]:
# Create new variables

In [49]:
# Define a function

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
          return 'High range product'
    else: return np.nan

In [50]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Lenad\AppData\Local\Temp\ipykernel_18680\93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [51]:
df['price_range'].value_counts(dropna = False)

price_range
Mid range product     673471
Low range product     314116
High range product     12413
Name: count, dtype: int64

In [52]:
df['prices'].max()

99999.0

In [53]:
# Define the function using loc()

In [54]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'
    

C:\Users\Lenad\AppData\Local\Temp\ipykernel_18680\736428746.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'


In [55]:
df.loc[(df['prices'] <=15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid range product'

In [56]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [57]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product     673471
Low range product     314116
High range product     12413
Name: count, dtype: int64

In [58]:
# Repeating process on entire dataframe

In [59]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High range product'

In [60]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid range product'

In [61]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [62]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product     21890146
Low range product     10126384
High range product      417682
Name: count, dtype: int64

In [63]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

In [64]:
# Find how busy different days are

result = []

for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [65]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [66]:
df_ords_prods_merge['busiest_day'] = result

In [109]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

In [68]:
# See new variables in dataframe

df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  8                    NaN       14084                  2   
2                  8                    NaN       12427                  3   
3                  8                    NaN       26088                  4   
4                  8                    NaN       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge    price_range_loc     busiest_day  
0              7     9.0   both  Mid range product  Regularly busy  
1             16    12.5   both  Mid range product  Regularly busy  
2             19     4.4   both  Low range product  Regularly busy  
3             19     4.7   both  Low range product  Regularly busy  
4             17     1.0   both  Low range product  Regularly busy

#### 4. Create busiest_days variable

'Busiest days' will refer to the two busiest days and 'Slowest days' will refer to the two slowest days. The three remaining days will be labelled 'Regularly busy'.

In [69]:
busiest_days_result = []

for value in df_ords_prods_merge['orders_day_of_week']:
  if value in range (0,2):
    busiest_days_result.append('Busiest days')
  elif value in range (3,5):
    busiest_days_result.append('Slowest days')
  else:
    busiest_days_result.append('Regularly busy')

In [70]:
busiest_days_result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

In [71]:
df_ords_prods_merge['busiest_days'] = busiest_days_result

In [72]:
# check values for accuracy

df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12927461
Busiest days      11875462
Slowest days       7631289
Name: count, dtype: int64

In [73]:
# Check values for accuracy - second check

The busiest days are Saturday (6209632 orders) and Sunday (5665830 orders)

In [74]:
a = 6209632
b = 5665830

In [75]:
sum = a + b

In [76]:
sum

11875462

The value for the busiest days is correct.

The slowest days are Tuesday (3844096 orders) and Wednesday (3787193 orders)

In [77]:
c = 3844096
d = 3787193

In [78]:
sum2 = c + d

In [79]:
sum2

7631289

The value for the slowest days is also correct.

#### 5. Create busiest_period_of_day variable

In [80]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

Looking at the value counts for the order_hour_of_day column, I think the best way of dividing the periods of time into 'most orders', 'average orders' and 'fewest orders' would be as follows:

Most orders - orders above or equal to 1,000,000. The applicable hours are between 0900 and 1700 hours.

Average orders - orders below 1,000,000 but above or equal to 400,000. The applicable hours are 0700, 0800 and between 1800 and 2300 hours.

Fewest orders - orders below 400,000. The applicable hours are between 0000 and 0600 hours

In [105]:
# Create busiest_period_of_day variable

busiest_period_of_day_result = []

for value in df_ords_prods_merge['order_hour_of_day']:
  if value in range (9,18):
    busiest_period_of_day_result.append('Most orders')
  elif value in range (0,7):
    busiest_period_of_day_result.append('Fewest orders')
  else:
    busiest_period_of_day_result.append('Average orders')

In [100]:
busiest_period_of_day_result

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [107]:
df_ords_prods_merge['busiest_period_of_day'] = busiest_period_of_day_result

In [111]:
# Check frequency for busiest_period_of_day column

df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       23226714
Average orders     8319912
Fewest orders       887586
Name: count, dtype: int64

In [113]:
# Check values for accuracy - second check

e = 290770
f = 218942
g = 115786
h = 88057
i = 69431
j = 53283
k = 51317

In [114]:
sum = e + f + g + h + i + j + k

In [115]:
sum

887586

887586 is the sum of all the orders placed between the hours of 0000 and 0600, which I have designated as 'Fewest orders'. This confirms that the new column is accurate.

In [116]:
# Check to see new column

df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  8                    NaN       14084                  2   
2                  8                    NaN       12427                  3   
3                  8                    NaN       26088                  4   
4                  8                    NaN       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge    price_range_loc     busiest_day  \
0              7     9.0   both  Mid range product  Regularly busy   
1             16    12.5   both  Mid range product  Regularly busy   
2             19     4.4   both  Low range product  Regularly busy   
3             19     4.7   both  Low range product  Regularly busy   
4             17     1.0   both  Low range product  Regularly busy   

     busiest_days busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy        Average orders  
3  Regularly busy        Average orders  
4  Regularly busy        Average orders

#### 6. Export data

In [118]:
# Export data to pkl

df_ords_prods_merge.to_pickle(os.path.join(path, '02. Data','Prepared Data', 'ords_prods_merge_derived.pkl'))